### PR TITLE
add xcode9 new signing support [JENKINS-45509]

### PR DESCRIPTION
**Pull request by lampietti**
[JENKINS-45509](https://issues.jenkins-ci.org/browse/JENKINS-45509)

1. wrap development team setup in a conditional manual signing block
2. add definitions of needed provisioning profiles in this same block
3. default to automatic signing with -allowProvisioningUpdates

### CODE REVIEW

**Comment by mesutdag** [comment]
Hi @lampietti 

I wonder if this PR solves the problem below , which occurs after upgrading XCode9

Error Domain=IDEFoundationErrorDomain Code=1 "exportOptionsPlist error for key 'iCloudContainerEnvironment': expected one of {Development, Production}, but no value was provided" UserInfo={NSLocalizedDescription=exportOptionsPlist error for key 'iCloudContainerEnvironment': expected one of {Development, Production}, but no value was provided}

It seems that we should also add 'iCloudContainerEnvironment' key to plist

**Comment by wens** [comment]
AFAIK you are still required to specify the development team even when using automatic signing if the project doesn't already specify one?
Otherwise I'd really like to see support for -allowProvisioningUpdates with export / packaging.
It would make management of auto signing provisioning profiles so much easier.

**Comment by lampietti** [comment]
@mesutdag it is not supported at this time, i am working on it

**Comment by lampietti** [comment]
@wens you're right, but if a team is already selected in the project, xcodebuild will fail if another development team is provided...
So, since it is a plugin for CI, i think we could expect developer using automatic signing have also selected a team.

**Comment by lampietti** [comment]
'iCloudContainerEnvironment' now supported with defaults values:  'Production' for 'app-store' export method, 'Development' for all others

**Comment by Glennzo** [comment]
I've been playing around with it, and what i noticed so far:

- teamID does not get set when picking a global defined Development Team, when setting the id directly in the target, everything is fine.

**Comment by yaogunfantuan** [comment]
when the pull requste will be agree to merge into master

**Comment by SwordJason** [comment]
I was kind of wait for this. When will this be merged or maybe someone could make a plugin that is working?

**Comment by mesutdag** [comment]
@SwordJason I have builded @lampietti 's PR and it is already working for us. You can download it here ; 
https://mylifebox.com/shr/011b26c0-53f9-4d1f-9fb7-009a3b39fe11&language=en

**Comment by lampietti** [comment]
@SwordJason you can download the HPI generated by the [CI build](https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fxcode-plugin/detail/PR-86/2/artifacts) for this PR

**Comment by SwordJason** [comment]
@lampietti @mesutdag Thank you very much.

**Comment by SwordJason** [comment]
I just got this information:
_Code Signing Error: No profile for team 'NB7MJZXXXX' matching 'Distribution_17' found:  Xcode couldn't find any provisioning profiles matching 'NB7MJZXXXX/Distribution_17'. Install the profile (by dragging and dropping it onto Xcode's dock item) or select a different one in the General tab of the target editor.
Code Signing Error: Code signing is required for product type 'Application' in SDK 'iOS 11.0'_

It's there something wrong configuration?

**Comment by kazuhidet** [comment]
In your modification method, the developmentTeamID is defined at line 615 of "XCodeBuilder.java" and the ID of the team acquired by "envs.expand (team.getTeamID ())" is substituted there, but this way , "developmentTeamID" does not reach around "// handle code signing identities" at line 764, a problem occurs if you select "Development Team" from the pull down in the setting "Code signing & OS X keychain options" .

So, I correct the problem and submitted pull request (PR#87).

**Comment by fbeeper** [comment]
Considering I haven't had much time and that I am not versed at all on the inner workings of this (or any) Jenkins plugin... from looking at the changes of this PR, I understand that the plugin would expect to get configured explicitly with the provisioning profiles. Otherwise, automatic signing would be used.

Xcode is asking for this provisioning profile config when exporting the IPA, and this implementation would honor that behavior. But wouldn't it be interesting to extract that from the build configuration so it is implicit at the job level (if defined in the config)?

I've created a simple bash script that uses xcodebuild to extract the provisioning profiles from the build settings while taking into account target and configuration variables. It's nothing fancy, I wrote it as a bash script to create my exportOptions.plist, but I guess could be easily translated into something else. Let me know if interested in working on this. I'd be glad to help to the extent of my knowledge 😄

**Comment by kazuhidet** [comment]
Although it is an experimental stage function which has not yet been fully tested, we have implemented a function to get information on the provisioning profile UUID, Bundle ID etc from the Xcode project file, which is required to do CodeSign.
This is useful when you want to compile existing code without taking time and effort.
(Especially if you do not want to update your profile or certificate without permission)

https://github.com/kazuhidet/xcode-plugin/tree/retrieve_codesign_info

File: src/main/java/au/com/rayh/XCodeBuilder.java [hunk 1]
```diff
@@ -540,16 +609,16 @@ private boolean _perform(Run<?,?> build, FilePath projectRoot, Launcher launcher
...
-        }
-        String developmentTeamID = envs.expand(team.getTeamID());
-        if (!StringUtils.isEmpty(developmentTeamID)) {
-            listener.getLogger().println(Messages.XCodeBuilder_DebugInfoCanFindPProfile());
-            /*returnCode =*/ launcher.launch().envs(envs).cmds("/usr/bin/security", "find-certificate", "-a", "-c", developmentTeamID, "-Z", "|", "grep", "^SHA-1").stdout(listener).pwd(projectRoot).join();
-            // We could fail here, but this doesn't seem to work as it should right now (output not properly redirected. We might need a parser)
+        if (team == null) {
+            listener.getLogger().println(Messages.XCodeBuilder_teamNotConfigured());
+        } else {
+            String developmentTeamID = envs.expand(team.getTeamID());
```
Comment (dimitryvolkov) [inline src/main/java/au/com/rayh/XCodeBuilder.java]: This change forces "developmentTeamID" into a reduced scope, causing it to appear as empty in the following lines, namely the lines exporting it to the ExportOptions.plist, causing the plugin to fail building...

**Comment by mathcoder23** [comment]
1,
### default to automatic signing with -allowProvisioningUpdates  
 why jenkins build not default  -allowProvisioningUpdates  ?  

2,
### my config  Custom xcodebuild arguments :  -allowProvisioningUpdates 
build archive hava `-allowProvisioningUpdates ` param ,but build ipa nothing

**Comment by devmedlmobile** [comment]
Would there happen to be an update on when this would be scheduled to be merged into master? Would definitely like to stick with the official release branch. For now, I will use the .hpi @lampietti linked above. Thank you @lampietti for building out this solution BTW! I am glad I searched the PR section here first before diving in myself lol.

**Pull request merged**